include/deal.II/multigrid/mg_smoother.h

```diff
@@ -997,13 +1094,86 @@ MGSmootherPrecondition<MatrixType, PreconditionerType, VectorType>::smooth
         {
           if (this->debug > 0)
             deallog << 'T';
-          if (i == 0 && u.all_zero())
-            *r = rhs;
-          else
-            {
-              matrices[level].Tvmult(*r,u);
-              r->sadd(-1.,1.,rhs);
-            }
+          matrices[level].Tvmult(*r,u);
```

Comment:
You are erasing the "shortcut" here because it turns out to be more expensive to call `u.all_zero()`, don't you?